### ISSUE TITLE
Istio egress gateway + TLS origination: 503 only for some domains

### ISSUE THREAD

**Issue by marcozov** [istio/istio#32907]
**Bug description**
I'm following https://istio.io/latest/docs/tasks/traffic-management/egress/egress-gateway-tls-origination/#perform-tls-origination-with-an-egress-gateway to perform TLS origination, in order to "convert" HTTP traffic directed to an external endpoint to HTTPS.
If I try the domain of the example (edition.cnn.com), things seem work pretty well. However, if I try other domains it's not really "deterministic": for some of them it works, for others it doesn't.
For instance: after setting up a new ServiceEntry, VirtualService, and DestinationRule (actually 2, one for sending traffic to the egress gateway and one for sending traffic to `www.facebook.com` from the egress gateway, as indicated in the documentation) for www.facebook.com, it works fine. But if I try something like www.gemuese.ch (after configuring the same set of resources for this domain) it does not:
```
root@command-demo-privileged-57c9c99d77-kdc4z:/# curl www.gemuese.ch
upstream connect error or disconnect/reset before headers. reset reason: connection failure
```
The access logs of the sidecar container of `command-demo-privileged-57c9c99d77-kdc4z`:
```
[2021-05-17T22:45:12.675Z] "GET / HTTP/1.1" 503 URX via_upstream - "-" 0 91 412 412 "-" "curl/7.64.0" "b3152ea2-5034-44c0-a86e-4567bb36b1dc" "www.gemuese.ch" "10.32.22.76:8080" outbound|80|gemuese|istio-egressgateway-http.istio-system.svc.cluster.local 10.32.18.17:50088 88.99.98.200:80 10.32.18.17:46026 - -
```
The access logs of the egress gateway:
```
[2021-05-17T23:02:00.311Z] "GET / HTTP/2" 503 UF,URX upstream_reset_before_response_started{connection_failure} - "-" 0 91 130 - "10.32.18.17" "curl/7.64.0" "886573f2-df9b-9682-b53d-71d626a61d67" "www.gemuese.ch" "88.99.98.200:443" outbound|443||www.gemuese.ch - 10.32.22.76:8080 10.32.18.17:43276 www.gemuese.ch -
[2021-05-17T23:02:00.452Z] "GET / HTTP/2" 503 UF,URX upstream_reset_before_response_started{connection_failure} - "-" 0 91 137 - "10.32.18.17" "curl/7.64.0" "886573f2-df9b-9682-b53d-71d626a61d67" "www.gemuese.ch" "88.99.98.200:443" outbound|443||www.gemuese.ch - 10.32.22.76:8080 10.32.18.17:43276 www.gemuese.ch -
[2021-05-17T23:02:00.630Z] "GET / HTTP/2" 503 UF,URX upstream_reset_before_response_started{connection_failure} - "-" 0 91 111 - "10.32.18.17" "curl/7.64.0" "886573f2-df9b-9682-b53d-71d626a61d67" "www.gemuese.ch" "88.99.98.200:443" outbound|443||www.gemuese.ch - 10.32.22.76:8080 10.32.18.17:43276 www.gemuese.ch -
```
If I enable debug logs on the egress gateway:
```
istioctl pc log --level "admin:debug,aws:debug,assert:debug,backtrace:debug,client:debug,config:debug,connection:debug,conn_handler:debug,dubbo:debug,file:debug,filter:debug,forward_proxy:debug,grpc:debug,hc:debug,health_checker:debug,http:debug,http2:debug,hystrix:debug,init:debug,io:debug,jwt:debug,main:debug,misc:debug,quic:debug,pool:debug,rbac:trace,router:debug,runtime:debug,stats:debug,secret:debug,tap:debug,testing:debug,tracing:debug,upstream:debug,udp:debug,envoy_bug:debug,ext_authz:debug,cache_filter:debug" istio-egressgateway-http-55c6c5d89d-fnvlh.istio-system

kubectl logs -n istio-system -f istio-egressgateway-http-55c6c5d89d-fnvlh --tail=100
...
...
2021-05-17T23:02:00.311946Z	debug	envoy pool	queueing stream due to no available connections
2021-05-17T23:02:00.311952Z	debug	envoy pool	creating a new connection
2021-05-17T23:02:00.336836Z	debug	envoy misc	Unknown error code 104 details Connection reset by peer
2021-05-17T23:02:00.336889Z	debug	envoy pool	[C10477] client disconnected, failure reason:
2021-05-17T23:02:00.336903Z	debug	envoy router	[C10231][S3091061571089400553] upstream reset: reset reason: connection failure, transport failure reason:
2021-05-17T23:02:00.356083Z	debug	envoy router	[C10231][S3091061571089400553] performing retry
2021-05-17T23:02:00.356132Z	debug	envoy pool	queueing stream due to no available connections
2021-05-17T23:02:00.356138Z	debug	envoy pool	creating a new connection
2021-05-17T23:02:00.380846Z	debug	envoy misc	Unknown error code 104 details Connection reset by peer
2021-05-17T23:02:00.380893Z	debug	envoy pool	[C10478] client disconnected, failure reason:
2021-05-17T23:02:00.380905Z	debug	envoy router	[C10231][S3091061571089400553] upstream reset: reset reason: connection failure, transport failure reason:
2021-05-17T23:02:00.417241Z	debug	envoy router	[C10231][S3091061571089400553] performing retry
2021-05-17T23:02:00.417303Z	debug	envoy pool	queueing stream due to no available connections
2021-05-17T23:02:00.417313Z	debug	envoy pool	creating a new connection
2021-05-17T23:02:00.442163Z	debug	envoy misc	Unknown error code 104 details Connection reset by peer
2021-05-17T23:02:00.442211Z	debug	envoy pool	[C10479] client disconnected, failure reason:
2021-05-17T23:02:00.442224Z	debug	envoy router	[C10231][S3091061571089400553] upstream reset: reset reason: connection failure, transport failure reason:
2021-05-17T23:02:00.442265Z	debug	envoy http	[C10231][S3091061571089400553] Sending local reply with details upstream_reset_before_response_started{connection failure}
2021-05-17T23:02:00.442317Z	debug	envoy http	[C10231][S3091061571089400553] encoding headers via codec (end_stream=false):
':status', '503'
'content-length', '91'
'content-type', 'text/plain'
'date', 'Mon, 17 May 2021 23:02:00 GMT'
'server', 'istio-envoy'

2021-05-17T23:02:00.442532Z	debug	envoy filter	Called AuthenticationFilter : onDestroy
2021-05-17T23:02:00.452372Z	debug	envoy http	[C10231] new stream
2021-05-17T23:02:00.452463Z	debug	envoy http	[C10231][S387142506190779831] request headers complete (end_stream=true):
':authority', 'www.gemuese.ch'
':path', '/'
':method', 'GET'
':scheme', 'https'
'user-agent', 'curl/7.64.0'
'accept', '*/*'
'x-forwarded-proto', 'http'
'x-request-id', '886573f2-df9b-9682-b53d-71d626a61d67'
'x-envoy-decorator-operation', 'istio-egressgateway-http.istio-system.svc.cluster.local:80/*'
```
I'm not sure how to debug that `Unknown error code 104 details Connection reset by peer`.
I don't get why a different URL gives this kind of error: as far as I understood, the target server just sees the HTTPS request. That's why I don't see why the behaviour should be different for different domains.

[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**
running `curl www.gemuese.ch` should return the same content as the https page.

**Steps to reproduce the bug**
Follow https://istio.io/latest/docs/tasks/traffic-management/egress/egress-gateway-tls-origination/#perform-tls-origination-with-an-egress-gateway but use another domain (e.g., www.gemuese.ch).

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)
```
istioctl  version --remote
client version: 1.9.0
control plane version: 1.9.2
data plane version: 1.9.2 (159 proxies)

kubectl version --short
Client Version: v1.19.3
Server Version: v1.19.7

helm version --short
v3.4.1+gc4e7485
```

**How was Istio installed?**
With the istio operator, via the helm chart on https://github.com/istio/istio.git (under the `manifests/charts/istio-operator` folder).

**Environment where the bug was observed (cloud vendor, OS, etc)**
AKS.

Additionally, please consider running `istioctl bug-report` and attach the generated cluster-state tarball to this issue.
Refer [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) for more details.

**Comment by ldemailly** [comment]
```
Not sure yet if I have the exact same issue but 
following https://istio.io/latest/docs/tasks/traffic-management/egress/egress-tls-origination/
I get

 503 UF,URX upstream_reset_before_response_started{connection_failure}

_only_ when doing tls origination to another istio 1.10.3 ingress (another cluster)


sslscan for the target shows
```
Version: 2.0.10-static
OpenSSL 1.1.1k  25 Mar 2021

Testing SSL server xxx on port 443 using SNI name xxx

  SSL/TLS Protocols:
SSLv2     disabled
SSLv3     disabled
TLSv1.0   enabled
TLSv1.1   enabled
TLSv1.2   enabled
TLSv1.3   enabled

  TLS Fallback SCSV:
Server supports TLS Fallback SCSV

  TLS renegotiation:
Secure session renegotiation supported

  TLS Compression:
OpenSSL version does not support compression
Rebuild with zlib1g-dev package for zlib support

  Heartbleed:
TLSv1.3 not vulnerable to heartbleed
TLSv1.2 not vulnerable to heartbleed
TLSv1.1 not vulnerable to heartbleed
TLSv1.0 not vulnerable to heartbleed

  Supported Server Cipher(s):
Preferred TLSv1.3  128 bits  TLS_AES_128_GCM_SHA256        Curve 25519 DHE 253
Accepted  TLSv1.3  256 bits  TLS_AES_256_GCM_SHA384        Curve 25519 DHE 253
Accepted  TLSv1.3  256 bits  TLS_CHACHA20_POLY1305_SHA256  Curve 25519 DHE 253
Preferred TLSv1.2  256 bits  ECDHE-RSA-CHACHA20-POLY1305   Curve 25519 DHE 253
Accepted  TLSv1.2  128 bits  ECDHE-RSA-AES128-GCM-SHA256   Curve 25519 DHE 253
Accepted  TLSv1.2  128 bits  ECDHE-RSA-AES128-SHA          Curve 25519 DHE 253
Accepted  TLSv1.2  128 bits  AES128-GCM-SHA256            
Accepted  TLSv1.2  128 bits  AES128-SHA                   
Accepted  TLSv1.2  256 bits  ECDHE-RSA-AES256-GCM-SHA384   Curve 25519 DHE 253
Accepted  TLSv1.2  256 bits  ECDHE-RSA-AES256-SHA          Curve 25519 DHE 253
Accepted  TLSv1.2  256 bits  AES256-GCM-SHA384            
Accepted  TLSv1.2  256 bits  AES256-SHA                   
Preferred TLSv1.1  128 bits  ECDHE-RSA-AES128-SHA          Curve 25519 DHE 253
Accepted  TLSv1.1  128 bits  AES128-SHA                   
Accepted  TLSv1.1  256 bits  ECDHE-RSA-AES256-SHA          Curve 25519 DHE 253
Accepted  TLSv1.1  256 bits  AES256-SHA                   
Preferred TLSv1.0  128 bits  ECDHE-RSA-AES128-SHA          Curve 25519 DHE 253
Accepted  TLSv1.0  128 bits  AES128-SHA                   
Accepted  TLSv1.0  256 bits  ECDHE-RSA-AES256-SHA          Curve 25519 DHE 253
Accepted  TLSv1.0  256 bits  AES256-SHA                   

  Server Key Exchange Group(s):
TLSv1.3  128 bits  secp256r1 (NIST P-256)
TLSv1.3  128 bits  x25519
TLSv1.2  128 bits  secp256r1 (NIST P-256)
TLSv1.2  128 bits  x25519

  SSL Certificate:
Signature Algorithm: sha256WithRSAEncryption
RSA Key Strength:    2048
```
it works fine for www.google.com for contrast

so I don't think it's not having tls1.2 (it's envoy talking to envoy and failing, for me)

**Comment by ldemailly** [comment]
I tried adding sni entry but that didn't help
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: echo-ext
spec:
  host: echo.dev.xxx
  trafficPolicy:
    portLevelSettings:
    - port:
        number: 80
      tls:
        mode: SIMPLE # initiates HTTPS when accessing echo.dev.xxx
        sni: echo.dev.xxx
```

**Comment by ldemailly** [comment]
Ok I found a working minimal solution:

source of traffic side car can be envoy from 1.10.3, doesn't need to be 1.11.3

`sni:` must be specified in DestinationRule trafficPolicy portLevelSettings tls

receiving ingress must be 1.11.0 (1.10.3 fails as above)

So my question is which bug fix in ingress/envoy fixed it between 1.10.3 and 1.11.0 can that be backported to 1.10.4?
(or is there a workaround in 1.10.3 ingress settings to make it work)

**Comment by ldemailly** [comment]
By the time I finished testing 1.11.0 and 1.10.3 matrix, 1.10.4 came out and.... it's fixed! 🎉 

So I would close this (but reopen if you still have issues after 1.10.4 and using `sni:` )
